wrappers/python/openmm/app/internal/xtc_utils/src/xtc.cpp

```diff
@@ -0,0 +1,326 @@
+    if (!natoms) {
+        fprintf(stderr, "xtc_read(): natoms is 0\n");
+        return;
+    }
+    xd = xdrfile_open(filename, "r");
+    if (NULL == xd) {
+        fprintf(stderr, "xtc_read(): could not open file\n");
+        return;
+    }
+    p = (rvec*)malloc(sizeof(rvec) * natoms);
+    int retval = 0;
+    int fidx = 0;
+    int _natoms_garbage = 0;
+    int xidx, yidx, zidx, aidx;
+    while (exdrOK == (retval = read_xtc(xd, _natoms_garbage, &step, &time, box, p, &prec))) {
+        time_arr[fidx] = time;
+        step_arr[fidx] = step;
+        for (int i = 0; i < 3; i++) {
+            for (int j = 0; j < 3; j++) {
+                box_arr[fidx + (3 * i + j) * nframes] = box[i][j];
+            }
+        }
+        for (aidx = 0; aidx < natoms; aidx++) {
+            xidx = Xf(aidx, fidx, nframes);
+            yidx = Yf(xidx, nframes);
+            zidx = Zf(yidx, nframes);
+            coords_arr[xidx] = p[aidx][0];
+            coords_arr[yidx] = p[aidx][1];
+            coords_arr[zidx] = p[aidx][2];
+        }
+        fidx++;
+    }
+    condfree((void*)p);
+    p = NULL;
+    xdrfile_close(xd);
+    if (retval == exdr3DX) {
+        fprintf(stderr, "xtc_read(): XTC file is corrupt\n");
+    }
+}
+
+struct XTC_frame* xtc_read(char* filename, int* natoms, int* nframes, double* dt, int* dstep) {
+    struct XTC_frame* frames = NULL;
+    *natoms = 0;
+    *nframes = 0;
+    rvec* p = NULL;
+    XDRFILE* xd = NULL;
+    float time;
+    int step;
+    float prec;
+    matrix box;
+    int i;
+    if (exdrOK != read_xtc_natoms(filename, natoms)) {
+        fprintf(stderr, "xtc_read(): could not get natoms\n");
+        return NULL;
+    }
+    if (!*natoms) {
+        fprintf(stderr, "xtc_read(): natoms is 0\n");
+        return NULL;
+    }
+    xd = xdrfile_open(filename, "r");
+    if (NULL == xd) {
+        fprintf(stderr, "xtc_read(): could not open file\n");
+        return NULL;
+    }
+    p = (rvec*)malloc(sizeof(rvec) * *natoms);
+    int retval = 0;
+    uint64_t offset = 0;
+    while (exdrOK == (retval = read_xtc(xd, *natoms, &step, &time, box, p, &prec))) {
+        frames = (struct XTC_frame*)realloc(frames, sizeof(struct XTC_frame) * (*nframes + 1));
+        if (!(frames)) {
+            fprintf(stderr, "xtc_read(): Allocation error in xtc.c (1). nframes=%d natoms=%d\n", *nframes, *natoms);
+            return NULL;
+        }
+        frames[*nframes].time = time;
+        frames[*nframes].step = step;
+        for (int i = 0; i < 3; i++) {
+            for (int j = 0; j < 3; j++) {
+                frames[*nframes].box[3 * i + j] = box[i][j];
+            }
+        }
+        frames[*nframes].pos = (float*)malloc(sizeof(float) * 3 * *natoms);
+        if (!(frames[*nframes].pos)) {
+            fprintf(stderr, "xtc_read(): Allocation error in xtc.c (2). nframes=%d natoms=%d\n", *nframes, *natoms);
+            return NULL;
```

Comment:
Here's a perfect example of what I was saying about memory allocation.  You just leaked the memory pointed to by `p`.  C++ has better ways of doing memory allocation that eliminate the possibility of errors like this.